### PR TITLE
Feat/dj adapter

### PR DESCRIPTION
**Pull request by HYLcool**
Support basic Adapter and Monitor:
- monitor resource utilization when running specified functions
- probe small data batch, monitor the resource utilization of each OP, and decide the batch size for each OP

Notice: datasets need to be updated to >= 2.19.0

- [ ] TODO: set batch size to 1 for all OPs before probing
- [ ] Wait for batch size support

### CODE REVIEW

File: data_juicer/core/adapter.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+                op.run, args=(dataset, ), sample_interval=sample_interval)
+
+            # calculate speed
+            resource_util_per_op[
+                'speed'] = sample_num / resource_util_per_op['time']
+            resource_util_list.append(resource_util_per_op)
+
+        return resource_util_list
+
+    def workloads_adapt(self, dataset, operators):
```
Comment (drcege) [inline data_juicer/core/adapter.py:45]: Looks like this is the main entrypoint; best to organize methods based on the call flow.

The name is somewhat unremarkable; might be `schedule/adapt_workloads/balance_workloads` to align with other naming styles.

File: data_juicer/core/adapter.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
+from data_juicer.core.monitor import Monitor
+
+
+class Adapter:
+
+    MAX_BATCH_SIZE = 10000
+
+    def __init__(self, cfg):
```
[the remark this answers is not in the document]
Comment (HYLcool) [reply data_juicer/core/adapter.py:8]: Done.

File: data_juicer/core/adapter.py (hunk 1 above)
Comment (HYLcool) [reply data_juicer/core/adapter.py:45]: Right, `adapt_workloads` seems better~

File: data_juicer/core/monitor.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+        Detect the resource utilization of all distributed nodes.
+        """
+        # TODO
+        raise NotImplementedError
+
+    @staticmethod
+    def monitor_current_resources():
+        """
+        Detect the resource utilization of the current environment/machine.
+        All data of "util." is in percent. All data of "mem." is in MB.
```
[the remark this answers is not in the document]
Comment (HYLcool) [reply data_juicer/core/monitor.py:92]: Yes. They were in percent before but changed to ratio later. Fixed.

File: data_juicer/core/adapter.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    def take_batch(dataset, config):
+        """
+        Split the dataset into batches based on configuration and load factor.
+
+        :param dataset: The dataset to be split
+        :param config: Configuration settings, including batch size
+        :return: An iterator of batches
+        """
+        # get initial batch size
+        batch_size = config.get('batch_size', 1000)
```
[the remark this answers is not in the document]
Comment (HYLcool) [reply data_juicer/core/adapter.py:128]: To align HF datasets library, we can set the default batch size to 1000. Here I changed it to DEFAULT_MAX_BATCH_SIZE from datasets.config, which can also be configurable.

File: data_juicer/core/adapter.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+class Adapter:
+
+    MAX_BATCH_SIZE = 10000
+
+    def __init__(self, cfg):
+        self.cfg = cfg
+        self.idle_resources = Monitor.monitor_current_resources()
+
+    @staticmethod
```
[the remark this answers is not in the document]
Comment (HYLcool) [reply data_juicer/core/adapter.py:14]: Done.

File: data_juicer/core/adapter.py [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+        for item in load_analysis_res:
+            max_util = 1e-5
+            max_key = min(left_utils.items(), key=lambda it: it[1])[0]
+            analysis_res = item['resource_analysis']
+            for key in analysis_res:
+                if 'util.' not in key or 'GPU' in key:
+                    continue
+                used_util = max(
+                    0, analysis_res[key]['max'] - self.idle_resources[key])
```
Comment (yxdyc) [inline data_juicer/core/adapter.py:138]: to discuss: using `max` or `avg`? As `max` is sensitive to the outlier within the probed small batch, we may have high probability to under-leverage the hardware resources.

File: data_juicer/core/adapter.py [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+        :return: A list of probe results for each OP.
+        """
+        # take a small batch
+        data_batch = self.take_batch(dataset, self.cfg)
+        # process and monitor the resource utilization
+        resource_util_list = self.execute_and_probe(data_batch, operators)
+        # analyze resource utilization
+        analysis_res = Monitor.analyze_resource_util_list(resource_util_list)
+
+        return analysis_res, len(data_batch)
```
Comment (yxdyc) [inline data_juicer/core/adapter.py:110]: In docstring "return: A list of probe results for each OP." --> "..., and the length of probed data batch

Comment (HYLcool) [reply data_juicer/core/adapter.py:111]: done.

File: data_juicer/core/adapter.py (hunk 6 above)
Comment (yxdyc) [reply data_juicer/core/adapter.py:139]: to be tested with more experiments and will incorporated in following PRs

**Review by yxdyc — APPROVED** [review]

**Pull request merged**